src/protocol/malicious.rs

```diff
@@ -157,64 +197,64 @@ impl<F: Field> SecurityValidator<F> {
     ///
     /// ## Panics
     /// Will panic if the mutex is poisoned
-    #[allow(clippy::await_holding_lock)]
-    pub async fn validate(self, ctx: SemiHonestContext<'_, F>) -> Result<(), Error> {
+    pub async fn validate<D: DowngradeMalicious>(self, values: D) -> Result<D::Target, Error> {
         // send our `u_i+1` value to the helper on the right
-        let channel = ctx.mesh();
-        let helper_right = ctx.role().peer(Direction::Right);
-        let helper_left = ctx.role().peer(Direction::Left);
-
-        let state = self.u_and_w.lock().unwrap();
-        try_join(
-            channel.send(helper_right, RECORD_0, state.u),
-            channel.send(helper_right, RECORD_1, state.w),
-        )
-        .await?;
-
-        // receive `u_i` value from helper to the left
-        let (u_left, w_left): (F, F) = try_join(
-            channel.receive(helper_left, RECORD_0),
-            channel.receive(helper_left, RECORD_1),
-        )
-        .await?;
-
-        let u_share = Replicated::new(u_left, state.u);
-        let w_share = Replicated::new(w_left, state.w);
+        let (u_share, w_share) = self.propagate_u_and_w().await?;
 
         // This should probably be done in parallel with the futures above
-        let r = ctx
-            .narrow(&Step::RevealR)
-            .reveal(RECORD_0, &self.r_share)
-            .await?;
+        let narrow_ctx = self.validate_ctx.narrow(&ValidateStep::RevealR);
+        let r = narrow_ctx.reveal(RECORD_0, &self.r_share).await?;
         let t = u_share - &(w_share * r);
 
-        let is_valid = check_zero(ctx.narrow(&Step::CheckZero), RECORD_0, &t).await?;
+        let check_zero_ctx = self.validate_ctx.narrow(&ValidateStep::CheckZero);
```

Comment:
It seems awfully wasteful to set up a new communication channel to only ever evaluate a single record. Instead of narrowing 3 times for these 3 steps (propagate u&w, reveal r, check zero) could we just use a single communication channel and make these different records?

Reply:
Maybe.  Later :)